src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
           }
         }
       }
+    } else {
+      // in passive materialization, we create a control node which covers multiple inputs of a region node.
+      //
+      // +-----------+   +--------+
+      // | CFG node  |   |cfg n   |
+      // +-----------+   +--------+
+      //      |  \  \     /
+      //      |   |  |   |
+      // +----------------+
+      // | Region         |
+      // +----------------+
+      //
+      // we need to normalize the region node, and its phi nodes.
+      // +-----------+  +-----+
+      // | CFG node  |  |cfg n|
+      // +-----------+  +-----+
+      //      |        /
+      //      |       |
+      // +-------------+
+      // | Region      | (inputs subtracts 2)
+      // +-------------+
+      //
+      for (uint i = 1; i < req()-1; i++) {
+        Node* c1 = in(i);
+        if (c1 == nullptr || c1->is_top()) continue;
+
+        GrowableArray<uint> paths;
+        for (uint j=i+1; j<req(); ++j) {
+          if (in(j) == c1) {
+            paths.push(j);
+          }
+        }
+
+        if (paths.length() > 0) {
+          PhaseIterGVN* igvn = phase->is_IterGVN();
+
+          for (DUIterator_Fast kmax, k = fast_outs(kmax); k < kmax; k++) {
+            Node* phi = fast_out(k);
+
+            if (phi != nullptr && phi->is_Phi() && phi->req() == req()) {
```

Comment:
Is `phi->req() == req()` always true? Phi's should have the same number of inputs as the region node.